### PR TITLE
Create dotty-staging library

### PR DESCRIPTION
**Pull request by nicolasstucki**
Extract the staging out of the standard library and compiler.
* Create `dotty-staging` project that depends on `dotty-compiler`
* Add missing test infrastructure dependencies on the `dotty-staging` project or jar
* Move `ScopeId` logic into `dotty.tools.dotc.quoted.QuoteContext`
* Remove `library/src-non-bootstrapped/scala/quoted/package.scala`
* Move `staging.{run, withQuoteContext}`, `Toolbox`, `QuoteDriver`, `QuoteCompiler`, `ExprCompilationUnit` to `dotty-staging`
* Implement `Toolbox.make` directly (instead of as a java reflection call)
* Now projects need to depend on `"ch.epfl.lamp" %% "dotty-staging" % scalaVersion.value`

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/consumetasty/ConsumeTasty.scala [hunk 1]
```diff
@@ -16,9 +18,23 @@ object ConsumeTasty {
...
     }
 
-    val currentClasspath = QuoteDriver.currentClasspath(getClass.getClassLoader)
+    val currentClasspath = getCurrentClasspath(getClass.getClassLoader)
     import java.io.File.{ pathSeparator => sep }
     val args = "-from-tasty" :: "-Yretain-trees" :: "-classpath" :: s"$classpath$sep$currentClasspath" :: classes
     (new Consume).process(args.toArray)
   }
+
+  private def getCurrentClasspath(cl: ClassLoader): String = {
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/consumetasty/ConsumeTasty.scala]: Why does this need to be duplicated with QuoteDriver ? (Unrelated but this implementation looks very fragile to me, this makes way too many assumptions about how classloaders are setup, which might be broken by sbt 1.3 uses of layered classloaders for example)

File: library/src/scala/quoted/ScopeException.scala [hunk 2]
```diff
@@ -0,0 +1,3 @@
+package scala.quoted
+
+class ScopeException(msg: String) extends Exception(msg)
```
Comment (smarter) [inline library/src/scala/quoted/ScopeException.scala]: Add some doc ?

File: project/Build.scala [hunk 3]
```diff
@@ -661,7 +661,8 @@ object Build {
           // running the compiler, we should always have the bootstrapped
           // library on the compiler classpath since the non-bootstrapped one
           // may not be binary-compatible.
-          "dotty-library"       -> packageBin.in(`dotty-library-bootstrapped`, Compile).value
+          "dotty-library"       -> packageBin.in(`dotty-library-bootstrapped`, Compile).value,
+          "dotty-staging"       -> packageBin.in(LocalProject("dotty-staging"), Compile).value,
```
Comment (smarter) [inline project/Build.scala]: dotty-staging should never be used with a non-bootstrapped compiler so I don't think this line is needed.

File: project/Build.scala [hunk 4]
```diff
@@ -738,6 +740,31 @@ object Build {
...
+    dependsOn(dottyCompiler(Bootstrapped)).
+    dependsOn(dottyCompiler(Bootstrapped) % "test->test").
+    settings(commonBootstrappedSettings).
+    settings(
+      fork in run := true,
+      fork in Test := true,
+      javaOptions ++= {
+        val attList = (dependencyClasspath in Runtime).value
+        val jars = packageAll.in(`dotty-compiler-bootstrapped`).value
+        List(
```
Comment (smarter) [inline project/Build.scala]: Too much duplication with the compiler javaOptions, either use the same javaOptions in both projects, or make a `def` to output the common parts.

**Comment by nicolasstucki** [comment]
Rebased

File: project/Build.scala (hunk 3 above)
Comment (nicolasstucki) [reply project/Build.scala]: removed it but had to adapt some code in the test infrastructure

File: library/src/scala/quoted/ScopeException.scala (hunk 2 above)
Comment (nicolasstucki) [reply library/src/scala/quoted/ScopeException.scala]: done

File: project/Build.scala (hunk 4 above)
Comment (nicolasstucki) [reply project/Build.scala]: I extracted in a common setting

File: compiler/src/dotty/tools/dotc/consumetasty/ConsumeTasty.scala (hunk 1 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/consumetasty/ConsumeTasty.scala]: I will look into it in another PR. For now we keep the same behaviour as before.

File: project/Build.scala [hunk 5]
```diff
@@ -746,6 +759,21 @@ object Build {
...
     )
 
+  lazy val `dotty-staging` = project.in(file("staging")).
+    withCommonSettings(Bootstrapped).
+    dependsOn(dottyCompiler(Bootstrapped)).
+    dependsOn(dottyCompiler(Bootstrapped) % "test->test").
+    settings(commonBootstrappedSettings).
+    settings(commonDottyJarClasspathSettings).
+    settings(
+      fork in run := true,
```
Comment (smarter) [inline project/Build.scala]: This line and the next should no longer be needed, this is now the default.

Comment (nicolasstucki) [reply project/Build.scala]: removed

File: project/Build.scala [hunk 6]
```diff
@@ -761,7 +761,11 @@ object Build {
 
   lazy val `dotty-staging` = project.in(file("staging")).
     withCommonSettings(Bootstrapped).
-    dependsOn(dottyCompiler(Bootstrapped)).
+    // We want the compiler to be present in the compiler classpath when compiling this project but not
+    // when compiling a project that depends on dotty-staging (see ...), but we always need it to be present
+    // on the JVM classpath at runtime.
+    dependsOn(dottyCompiler(Bootstrapped) % "provided").
+    dependsOn(dottyCompiler(Bootstrapped) % "compile->runtime").
```
Comment (nicolasstucki) [inline project/Build.scala]: @smarter I had to separate the dependencies. `% "provided,compile->runtime,test->test"` was not working.

**Pull request merged**